codex-rs/exec-server/src/posix.rs

```diff
@@ -0,0 +1,627 @@
+    pub exit_code: i32,
+    pub output: String,
+    pub duration: Duration,
+    pub timed_out: bool,
+}
+
+fn decide_escalate(file: &str, argv: &[String], _workdir: &PathBuf) -> EscalateAction {
+    // TODO: execpolicy
+    match (file, argv) {
+        ("/opt/homebrew/bin/gh", [_, arg1, arg2, ..]) if arg1 == "issue" && arg2 == "list" => {
+            EscalateAction::Escalate
+        }
+        _ => EscalateAction::RunInSandbox,
+    }
+}
+
+async fn handle_escalate_session_with_decider<F>(
+    socket: AsyncSocket,
+    decider: F,
+) -> anyhow::Result<()>
+where
+    F: Fn(&str, &[String], &PathBuf) -> EscalateAction,
+{
+    let EscalateClientMessage::EscalateRequest {
+        file,
+        argv,
+        workdir,
+        env,
+    } = socket.receive::<EscalateClientMessage>().await?;
+    let action = decider(&file, &argv, &workdir);
+    tracing::debug!("decided {action:?} for {file:?} {argv:?} {workdir:?}");
+    match action {
+        EscalateAction::RunInSandbox => {
+            socket
+                .send(EscalateServerMessage::EscalateResponse(
+                    EscalateAction::RunInSandbox,
+                ))
+                .await?;
+        }
+        EscalateAction::Escalate => {
+            socket
+                .send(EscalateServerMessage::EscalateResponse(
+                    EscalateAction::Escalate,
+                ))
+                .await?;
+            let (msg, fds) = socket
+                .receive_with_fds::<SuperExecMessage>()
+                .await
+                .context("failed to receive SuperExecMessage")?;
+            if fds.len() != msg.fds.len() {
+                return Err(anyhow::anyhow!(
+                    "mismatched number of fds in SuperExecMessage: {} in the message, {} from the control message",
+                    msg.fds.len(),
+                    fds.len()
+                ));
+            }
+
+            if msg
+                .fds
+                .iter()
+                .any(|src_fd| fds.iter().any(|dst_fd| dst_fd.as_raw_fd() == *src_fd))
+            {
+                return Err(anyhow::anyhow!(
+                    "overlapping fds not yet supported in SuperExecMessage"
+                ));
+            }
+
+            let mut command = Command::new(file);
+            command
+                .args(&argv[1..])
+                .arg0(argv[0].clone())
+                .envs(&env)
+                .current_dir(&workdir)
+                .stdin(Stdio::null())
+                .stdout(Stdio::null())
+                .stderr(Stdio::null());
+            unsafe {
+                command.pre_exec(move || {
+                    for (dst_fd, src_fd) in msg.fds.iter().zip(&fds) {
+                        libc::dup2(src_fd.as_raw_fd(), *dst_fd);
+                    }
+                    Ok(())
+                });
+            }
+            let mut child = command.spawn()?;
+            let exit_status = child.wait().await?;
+            socket
```

Comment:
I suppose we don't enforce a timeout because we rely on the invoker of the original tool call to enforce that? Though will we end up with stray processes hanging around if a timeout happens?

Will killing Bash make it through to here and ensure this is killed, as well?